File: 813.最大平均值和的分组.py
#
# @lc app=leetcode.cn id=813 lang=python3
#
# [813] 最大平均值和的分组
#

# @lc code=start
from typing import List


class Solution:
    def largestSumOfAverages(self, nums: List[int], k: int) -> float:
        n = len(nums)
        presum = [0] * (n + 1)
        for i, num in enumerate(nums):
            presum[i + 1] = presum[i] + num
        dp = [[0 for _ in range(k + 1)] for _ in range(n)]
        for i in range(n):
            dp[i][1] = presum[i + 1] / (i + 1)
        
        for j in range(2, k + 1):
            for i in range(j - 1, n):
                ans = 0
                for step in range(j - 2, i):
                    ans = max(ans, dp[step][j - 1] + (presum[i + 1] - presum[step + 1]) / (i - step))
                dp[i][j] = ans
        return dp[-1][-1]
# @lc code=end

File: 813.最大平均值和的分组.cpp
/*
 * @lc app=leetcode.cn id=813 lang=cpp
 *
 * [813] 最大平均值和的分组
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    double largestSumOfAverages(vector<int>& nums, int K) {
        int n = nums.size(), presum[n + 1];
        memset(presum, 0, sizeof(presum));
        vector<vector<double>> dp(n, vector<double>(K + 1, 0.0));

        for (int i = 0; i < n; i++) {
            presum[i + 1] = presum[i] + nums[i];
            dp[i][1] = presum[i + 1] * 1.0 / (i + 1);
        }
        for (int k = 2; k <= K; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = k - 2; j < i; j++) {
                    dp[i][k] = max(dp[i][k], dp[j][k - 1] + (presum[i + 1] - presum[j + 1]) * 1.0 / (i - j));
                }
            }
        }

        return dp[n - 1][K];
    }
};
// @lc code=end

int main(int argc, char const *argv[])
{
    Solution solution = Solution();
    vector<int> nums = {1,2,3,4,5,6,7};
    int k = 4;
    cout << solution.largestSumOfAverages(nums, k) << endl;
    return 0;
}
